design/cluster-api-provider-metal3/configurable_disk_cleaning.md

```diff
@@ -0,0 +1,234 @@
+will update `disableAutomatedClean` field value on all the respective
+Metal3Machines referenced by that Metal3MachineTemplate.
+
+##### Metal3Machine object
+
+We introduce a new field `disableAutomatedClean` in the spec of the
+Metal3Machine object. `disableAutomatedClean` can store a boolean type
+of a value. Setting it to True means do not perform automated
+cleaning, while False value enables automated cleaning. By default
+`disableAutomatedClean` will be set to False (i.e. do perform automated
+cleaning as it is now).
+
+```yaml
+apiVersion: infrastructure.cluster.x-k8s.io/v1alpha4
+kind: Metal3Machine
+metadata:
+spec:
+  template:
+  spec:
+   disableAutomatedClean: False   #default value. Perform automated cleaning.
+
+```
+
+#### Baremetal Operator
+
+We introduce a new field `disableAutomatedClean` in the spec of the
+Baremetalhost object. `disableAutomatedClean` can store a boolean type
+of a value. If set to True, Baremetal Operator instructs Ironic
+to disable automated cleaning for an Ironic node, while if
+set to False, Baremetal Operator instructs Ironic to enable automated
+cleaning. By default `disableAutomatedClean` will be set to False
+(i.e. do perform automated cleaning as it is now).
+
+```yaml
+apiVersion: metal3.io/v1alpha1
+kind: BareMetalHost
+metadata:
+  name: worker-0
+spec:
+  disableAutomatedClean: False   #default value. Perform automated cleaning.
+  ...
+```
+
+## Design Details
+
+**Example flow:**
+
+1. User wants to disable disk cleaning before upgrading the nodes that are part
+    of the same KCP/MD.
+
+2. User sets True for the `disableAutomatedClean` field on a
+    Metal3MachineTemplate which is referenced by `infrastructureTemplate` field
+    in the KCP/MD.
+
+3. Metal3MachineTemplate controller keeps reconciling the Metal3MachineTemplate objects.
+    Once the update is seen on the Metal3MachineTemplate, Metal3MachineTemplate controller
+    starts mapping all the Meta3Machines referenced by that particular Metla3MachineTemplate,
+    and updates the `disableAutomatedClean` field to True on all the referenced Metal3Machines.
+
+4. Metal3Machine controller keeps reconciling the Metal3Machine objects.
+    Once the update is seen on the Metal3Machine, Metal3Machine controller starts
+    mapping the BareMetalHosts referenced by the Metal3Machines, and updates the
+    `disableAutomatedClean` field to True on all the referenced BaremetalHosts.
+
+5. Since the `disableAutomatedClean` field is set to True on the BaremetalHost,
+    Baremetal Operator instructs Ironic to disable disk cleaning for the Ironic
+    nodes referenced by the BaremetalHosts.
+
+### Implementation Details/Notes/Constraints
+
+- As a user, you should not modify `disableAutomatedClean` field directly
+    on a Metal3Machine or BareMetalHost but rather do it through
+    Metal3MachineTemplate so that you can update the field on all the Metal3Machines
+    referenced by a single Metal3MachineTemplate at once.
```

Comment:
This feels limiting. If I deploy a bunch of hosts but want to pull one out of the cluster without erasing it for some sort of maintenance, shouldn't I be able to do that?

Reply:
It is a limitation exactly as you said and we discussed it internally with @maelk. The reason why we want to control it through Metal3MachineTemplate is to be able to successfully complete the upgrade operation.

In other words, we need to put a logic in Metal3 to know - to differentiate which nodes will be affected by the upgrade that user is about to start. So, normally that happens for a group of nodes under the same KCP/MD. 

And, we thought if we ask a user to express which nodes should have its cleaning disabled/enabled via Metal3MachineTemplate, then in Metal3 we would take care of passing that all the way down to Ironic nodes. 

User updates disableAutomatedClean on Metal3MachineTemplate
CAPM3 controllers will update Metal3Machines -> BareMetalHosts

So, Metal3MachineTemplate was the only solution to select nodes referenced by the same KCP/MD. And, as you can see the proposal is really focused on the upgrade.

If we update `disableAutomatedClean` on a BareMetalHost, then it is problematic, because - 

If that BareMetalHost is part of a KCP/MD that has more than one Machines/Metal3Machines/BareMetalHosts, then we will end up having N number of BMHs with different values in `disableAutomatedClean`. This is okay, unless you are not planning to perform the upgrade.
Even if the user updates disableAutomatedClean on a BareMetalHosts, we can’t simply update top layer objects (Metal3Machine, Metal3MachineTemplate) based on this change.  Because, there might be more than one BMH under the same KCP/MD (or Metal3MachineTemplate) and we can’t force disableAutomatedClean to be the same on others from bottom to top.

Of course, we can update `disableAutomatedClean` on a bmh, and BMO will communicate with Ironic to have the same value in the corresponding Ironic node. But if you think about what will happen when the user starts the upgrading - then I would say don’t modify it via BMH. 

But I would really appreciate it if you have some ideas/feedback to share on what we should do at this point. Should we make it clear that the user should not modify it via a BareMetalHost/Metal3Machine or still allow it but then what would we do when someone starts the upgrade?

One thing we can do, allow users to update that field on a BareMetalHost/Metal3Machine. And, if the user does the upgrade unexpectedly - CAPM3  will find  the Metal3Machines, and BareMetalHosts references by the same Metal3MachineTemplate and whatever value is set in Metal3MachineTemplate will be forcefully updated to be the same on all the BareMetalHosts. As such, if it’s not an upgrade, we are fine, and we don’t have to worry about updating Metal3Machine or Metal3MachineTemplate.

Reply:
> It is a limitation exactly as you said and we discussed it internally with @maelk. The reason why we want to control it through Metal3MachineTemplate is to be able to successfully complete the upgrade operation.
> 
> In other words, we need to put a logic in Metal3 to know - to differentiate which nodes will be affected by the upgrade that user is about to start. So, normally that happens for a group of nodes under the same KCP/MD.
> 
> And, we thought if we ask a user to express which nodes should have its cleaning disabled/enabled via Metal3MachineTemplate, then in Metal3 we would take care of passing that all the way down to Ironic nodes.

If the template's version of `disableAutomatedClean` is a `*bool` then we can tell the difference between `false` and not set. When the value in the template is not set, CAPM3 doesn't need to update the machine. That would allow someone to modify a single machine independently, while still supporting the "top down" propagation described in the algorithm here for the upgrade case.

Reply:
I agree with Doug here. I think that whenever the value is set in the template, it should take precedence to what the user has configured directly on the BMH and overwrite it during the reconciliation. If unset, let's leave it up to the user, agreed.

Reply:
Updated the document. Thanks @dhellmann , @maelk